{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "\n",
    "from argparse import Namespace\n",
    "import pbiotools.misc.logging_utils as logging_utils\n",
    "\n",
    "args = Namespace()\n",
    "logger = logging_utils.get_ipython_logger()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as mtick\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set(style=\"white\", color_codes=True, palette=\"muted\")\n",
    "\n",
    "import pbiotools.misc.mpl_utils as mpl_utils\n",
    "\n",
    "import yaml\n",
    "\n",
    "import pbiotools.ribo.ribo_utils as ribo_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the keys do not matter, but they must match in all dictionaries\n",
    "\n",
    "config_files = {\"my_ribo_project\": \"/path/to/my/config.yaml\"}\n",
    "\n",
    "# the counts produced with get-all-read-filtering-counts\n",
    "alignment_counts_files = {\"my_ribo_project\": \"/path/to/my/read-filtering-counts.csv.gz\"}\n",
    "\n",
    "# this file will be used for the output if args.without_rrna is False\n",
    "out_files = {\"my_ribo_project\": \"/path/to/my/read-filtering-counts.pdf\"}  # or png\n",
    "\n",
    "# this file will be used for the output if args.without_rrna is True\n",
    "without_rrna_files = {\n",
    "    \"my_ribo_project\": \"/path/to/my/read-filtering-counts.no-rrna.pdf\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this must match the dictionaries\n",
    "data = \"my_ribo_project\"\n",
    "\n",
    "args.without_rrna = False\n",
    "\n",
    "args.fontsize = 20\n",
    "args.legend_fontsize = 15\n",
    "\n",
    "if args.without_rrna:\n",
    "    args.ymax = 2e7 + 1\n",
    "    args.ystep = 5e6\n",
    "else:\n",
    "    args.ymax = 1.5e8 + 1\n",
    "    args.ystep = 2.5e7\n",
    "\n",
    "\n",
    "args.alignment_counts_order = [\n",
    "    \"raw_data_count\",\n",
    "    \"without_adapters_count\",\n",
    "    \"without_rrna_count\",\n",
    "    \"genome_count\",\n",
    "    \"unique_count\",\n",
    "    \"length_count\",\n",
    "]\n",
    "\n",
    "args.alignment_counts_names = [\n",
    "    \"Poor quality\",\n",
    "    \"Ribosomal\",\n",
    "    \"No alignment\",\n",
    "    \"Multimappers\",\n",
    "    \"Non-periodic\",\n",
    "    \"Usable\",\n",
    "]\n",
    "\n",
    "args.without_rrna_order = [\n",
    "    \"without_rrna_count\",\n",
    "    \"genome_count\",\n",
    "    \"unique_count\",\n",
    "    \"length_count\",\n",
    "]\n",
    "\n",
    "args.without_rrna_names = [\"No alignment\", \"Multimappers\", \"Non-periodic\", \"Usable\"]\n",
    "\n",
    "if args.without_rrna:\n",
    "    args.alignment_counts_order = args.without_rrna_order\n",
    "    args.alignment_counts_names = args.without_rrna_names\n",
    "\n",
    "\n",
    "args.config = config_files[data]\n",
    "args.alignment_counts = alignment_counts_files[data]\n",
    "\n",
    "\n",
    "args.out = out_files[data]\n",
    "if args.without_rrna:\n",
    "    args.out = without_rrna_files[data]\n",
    "\n",
    "args.alignment_counts_order = args.alignment_counts_order[::-1]\n",
    "args.alignment_counts_names = args.alignment_counts_names[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "msg = \"Reading config file\"\n",
    "logger.info(msg)\n",
    "\n",
    "config = yaml.load(open(args.config))\n",
    "sample_name_map = ribo_utils.get_sample_name_map(config)\n",
    "\n",
    "msg = \"Reading counts\"\n",
    "logger.info(msg)\n",
    "\n",
    "alignment_counts = pd.read_csv(args.alignment_counts)\n",
    "alignment_counts = alignment_counts.sort_values(\"note\").reset_index()\n",
    "\n",
    "names = alignment_counts[\"note\"]\n",
    "\n",
    "alignment_diff_counts = mpl_utils.get_diff_counts(\n",
    "    alignment_counts[args.alignment_counts_order]\n",
    ")\n",
    "df = pd.DataFrame(alignment_diff_counts)\n",
    "df.columns = args.alignment_counts_names\n",
    "df[\"name\"] = names\n",
    "\n",
    "# have to use apply rather than map because of weird interaction\n",
    "# with the missing_key_dictionary\n",
    "df[\"display_name\"] = df[\"name\"].apply(lambda x: sample_name_map[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "pal = sns.palettes.color_palette(\n",
    "    palette=\"Set3\", n_colors=len(args.alignment_counts_names)\n",
    ")\n",
    "\n",
    "gap = 0.15\n",
    "yticks = np.arange(0, args.ymax, args.ystep)\n",
    "\n",
    "bars = mpl_utils.create_stacked_bar_graph(\n",
    "    ax,\n",
    "    alignment_diff_counts,\n",
    "    colors=pal,\n",
    "    x_tick_labels=df[\"display_name\"],\n",
    "    y_ticks=yticks,\n",
    "    y_tick_labels=yticks,\n",
    "    gap=gap,\n",
    "    end_gaps=True,\n",
    "    stack_labels=args.alignment_counts_names,\n",
    "    y_title=\"Reads\",\n",
    "    log=False,\n",
    "    font_size=args.fontsize,\n",
    "    edge_colors=\"0.5\",\n",
    ")\n",
    "\n",
    "ax.legend(\n",
    "    loc=\"upper center\",\n",
    "    bbox_to_anchor=(0.5, -0.5),\n",
    "    ncol=3,\n",
    "    fontsize=args.legend_fontsize,\n",
    "    title=\"Filter\",\n",
    "    frameon=True,\n",
    "    framealpha=0.9,\n",
    ")\n",
    "\n",
    "if args.without_rrna:\n",
    "    ax.yaxis.set_major_formatter(mtick.FormatStrFormatter(\"%.0e\"))\n",
    "else:\n",
    "    ax.yaxis.set_major_formatter(mtick.FormatStrFormatter(\"%.0e\"))\n",
    "\n",
    "mpl_utils.set_label_fontsize(ax, args.fontsize)\n",
    "mpl_utils.set_legend_title_fontsize(ax, args.fontsize)\n",
    "\n",
    "if args.out is not None:\n",
    "    # fig.tight_layout()\n",
    "    fig.savefig(args.out, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
